### PR TITLE
Add cryptarchia consensus service

### PR DESCRIPTION
**Pull request by zeegomo**
First PR for adding cryptarchia support in nomos node.

Leadership and tests will come in following iterations

### CODE REVIEW

File: nomos-services/cryptarchia-consensus/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    consensus: cryptarchia_engine::Cryptarchia<HeaderId>,
+}
+
+impl Cryptarchia {
+    fn tip(&self) -> HeaderId {
+        self.consensus.tip()
+    }
+
+    fn try_apply_header(&self, header: &Header) -> Result<Self, Error> {
+        let header = header.cryptarchia();
```
Comment (davidrusu) [inline nomos-services/cryptarchia-consensus/src/lib.rs:50]: Why not simply take `cryptarchia::Header` here and force the caller to prove an instance of `cryptarchia::Header`?

This seems falsely general (i.e. type sig claims to work for any header)

File: nomos-services/cryptarchia-consensus/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        let id = header.id();
+        let parent = header.parent();
+        let slot = header.slot();
+        let ledger = self.ledger.try_update(
+            id,
+            parent,
+            header.slot(),
```
Comment (davidrusu) [inline nomos-services/cryptarchia-consensus/src/lib.rs:57]: `header.slot()` is stored in the `slot` variable above.

Is header moved in this call to `try_update()`? If not, you could inline all the variables (id()|parent()|slot())

**Review by youngjoon-lee — APPROVED** [review]
Looks good. I'll open a PR for the mixnet network adapter for the cryptarchia consensus service.

File: nomos-services/cryptarchia-consensus/src/lib.rs (hunk 1 above)
Comment (zeegomo) [reply nomos-services/cryptarchia-consensus/src/lib.rs:50]: fair enough

**Review by davidrusu — APPROVED** [review]

**Pull request merged**